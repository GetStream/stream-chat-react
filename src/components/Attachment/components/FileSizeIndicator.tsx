import React from 'react';
import prettybytes from 'pretty-bytes';

type FileSizeIndicatorProps = {
  /** file size in byte */
  fileSize?: number;
  /**
   The maximum number of fraction digits to display. If not set, the default behavior is to round to 3 significant digits.
   @default undefined
   */
  maximumFractionDigits?: number;
};

export const FileSizeIndicator = ({ fileSize, maximumFractionDigits }: FileSizeIndicatorProps) => {
  if (!(fileSize && Number.isFinite(Number(fileSize)))) return null;

  return (
    <span
      className='str-chat__message-attachment-file--item-size'
      data-testid='file-size-indicator'
    >
      {prettybytes(fileSize, { maximumFractionDigits })}
    </span>
  );
};
