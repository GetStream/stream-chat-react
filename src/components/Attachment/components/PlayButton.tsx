import React from 'react';
import { PauseIcon, PlayTriangleIcon } from '../icons';

type PlayButtonProps = {
  isPlaying: boolean;
  onClick: () => void;
};

export const PlayButton = ({ isPlaying, onClick }: PlayButtonProps) => (
  <button
    className='str-chat__message-attachment-audio-widget--play-button'
    data-testid={isPlaying ? 'pause-audio' : 'play-audio'}
    onClick={onClick}
  >
    {isPlaying ? <PauseIcon /> : <PlayTriangleIcon />}
  </button>
);
