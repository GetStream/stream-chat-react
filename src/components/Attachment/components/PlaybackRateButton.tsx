import React from 'react';

export type PlaybackRateButtonProps = React.ComponentProps<'button'>;

export const PlaybackRateButton = ({ children, onClick }: PlaybackRateButtonProps) => (
  <button
    className='str-chat__message_attachment__playback-rate-button'
    data-testid='playback-rate-button'
    onClick={onClick}
  >
    {children}
  </button>
);
