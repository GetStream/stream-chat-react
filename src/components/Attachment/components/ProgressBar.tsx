import React from 'react';

export type ProgressBarProps = {
  /** Progress expressed in fractional number value btw 0 and 100. */
  progress: number;
} & Pick<React.ComponentProps<'div'>, 'onClick'>;

export const ProgressBar = ({ onClick, progress }: ProgressBarProps) => (
  <div
    className='str-chat__message-attachment-audio-widget--progress-track'
    data-progress={progress}
    data-testid='audio-progress'
    onClick={onClick}
    role='progressbar'
    style={
      {
        '--str-chat__message-attachment-audio-widget-progress': progress + '%',
      } as React.CSSProperties
    }
  >
    <div
      className='str-chat__message-attachment-audio-widget--progress-slider'
      style={{ left: `${progress}px` }}
    />
  </div>
);
