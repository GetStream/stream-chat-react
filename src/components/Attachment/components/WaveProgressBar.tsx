import React, { MouseEventHandler, useMemo, useRef, useState } from 'react';
import clsx from 'clsx';
import { divMod } from '../utils';

type WaveProgressBarProps = {
  /** Function that allows to change the track progress */
  seek: MouseEventHandler<HTMLDivElement>;
  /** The array of fractional number values between 0 and 1 representing the height of amplitudes */
  waveformData: number[];
  /** Allows to specify the number of bars into which the original waveformData array should be resampled */
  amplitudesCount?: number;
  /** Progress expressed in fractional number value btw 0 and 100. */
  progress?: number;
};
export const WaveProgressBar = ({
  amplitudesCount = 40,
  progress = 0,
  seek,
  waveformData,
}: WaveProgressBarProps) => {
  const [progressIndicator, setProgressIndicator] = useState<HTMLDivElement | null>(null);
  const isDragging = useRef(false);

  const handleMouseDown = () => {
    if (!progressIndicator) return;
    isDragging.current = true;
    progressIndicator.style.cursor = 'grabbing';
  };

  const handleMouseMove: MouseEventHandler<HTMLDivElement> = (event) => {
    if (!isDragging.current) return;
    seek(event);
  };

  const handleMouseUp = () => {
    if (!progressIndicator) return;
    isDragging.current = false;
    progressIndicator.style.removeProperty('cursor');
  };

  const resampledWaveformData = useMemo(
    () =>
      waveformData.length === amplitudesCount
        ? waveformData
        : waveformData.length > amplitudesCount
        ? downSample(waveformData, amplitudesCount)
        : upSample(waveformData, amplitudesCount),
    [amplitudesCount, waveformData],
  );

  if (!waveformData.length) return null;

  return (
    <div
      className='str-chat__wave-progress-bar__track'
      data-testid='wave-progress-bar-track'
      onClick={seek}
      onMouseDown={handleMouseDown}
      onMouseLeave={handleMouseUp}
      onMouseMove={handleMouseMove}
      onMouseUp={handleMouseUp}
      role='progressbar'
    >
      {resampledWaveformData.map((amplitude, i) => (
        <div
          className={clsx('str-chat__wave-progress-bar__amplitude-bar', {
            ['str-chat__wave-progress-bar__amplitude-bar--active']:
              progress > (i / resampledWaveformData.length) * 100,
          })}
          data-testid='amplitude-bar'
          key={`amplitude-${i}`}
          style={
            {
              '--str-chat__wave-progress-bar__amplitude-bar-height': amplitude
                ? amplitude * 100 + '%'
                : '0%',
            } as React.CSSProperties
          }
        />
      ))}
      <div
        className='str-chat__wave-progress-bar__progress-indicator'
        data-testid='wave-progress-bar-progress-indicator'
        ref={setProgressIndicator}
        style={{ left: `${progress < 0 ? 0 : progress > 100 ? 100 : progress}%` }}
      />
    </div>
  );
};

/**
 * The downSample function uses the Largest-Triangle-Three-Buckets (LTTB) algorithm.
 * See the thesis Downsampling Time Series for Visual Representation by Sveinn Steinarsson for more (https://skemman.is/bitstream/1946/15343/3/SS_MSthesis.pdf)
 * @param data
 * @param targetOutputSize
 */
export function downSample(data: number[], targetOutputSize: number): number[] {
  if (data.length <= targetOutputSize || targetOutputSize === 0) {
    return data;
  }

  if (targetOutputSize === 1) return [mean(data)];

  const result: number[] = [];
  // bucket size adjusted due to the fact that the first and the last item in the original data array is kept in target output
  const bucketSize = (data.length - 2) / (targetOutputSize - 2);
  let lastSelectedPointIndex = 0;
  result.push(data[lastSelectedPointIndex]); // Always add the first point
  let maxAreaPoint, maxArea, triangleArea;

  for (let bucketIndex = 1; bucketIndex < targetOutputSize - 1; bucketIndex++) {
    const previousBucketRefPoint = data[lastSelectedPointIndex];
    const nextBucketMean = getNextBucketMean(data, bucketIndex, bucketSize);

    const currentBucketStartIndex = Math.floor((bucketIndex - 1) * bucketSize) + 1;
    const nextBucketStartIndex = Math.floor(bucketIndex * bucketSize) + 1;
    const countUnitsBetweenAtoC = 1 + nextBucketStartIndex - currentBucketStartIndex;

    maxArea = triangleArea = -1;

    for (
      let currentPointIndex = currentBucketStartIndex;
      currentPointIndex < nextBucketStartIndex;
      currentPointIndex++
    ) {
      const countUnitsBetweenAtoB = Math.abs(currentPointIndex - currentBucketStartIndex) + 1;
      const countUnitsBetweenBtoC = countUnitsBetweenAtoC - countUnitsBetweenAtoB;
      const currentPointValue = data[currentPointIndex];

      triangleArea = triangleAreaHeron(
        triangleBase(Math.abs(previousBucketRefPoint - currentPointValue), countUnitsBetweenAtoB),
        triangleBase(Math.abs(currentPointValue - nextBucketMean), countUnitsBetweenBtoC),
        triangleBase(Math.abs(previousBucketRefPoint - nextBucketMean), countUnitsBetweenAtoC),
      );

      if (triangleArea > maxArea) {
        maxArea = triangleArea;
        maxAreaPoint = data[currentPointIndex];
        lastSelectedPointIndex = currentPointIndex;
      }
    }

    if (typeof maxAreaPoint !== 'undefined') result.push(maxAreaPoint);
  }

  result.push(data[data.length - 1]); // Always add the last point

  return result;
}

const triangleAreaHeron = (a: number, b: number, c: number) => {
  const s = (a + b + c) / 2;
  return Math.sqrt(s * (s - a) * (s - b) * (s - c));
};

const triangleBase = (a: number, b: number) => Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));

const mean = (values: number[]) => values.reduce((acc, value) => acc + value, 0) / values.length;

const getNextBucketMean = (data: number[], currentBucketIndex: number, bucketSize: number) => {
  const nextBucketStartIndex = Math.floor(currentBucketIndex * bucketSize) + 1;
  let nextNextBucketStartIndex = Math.floor((currentBucketIndex + 1) * bucketSize) + 1;
  nextNextBucketStartIndex =
    nextNextBucketStartIndex < data.length ? nextNextBucketStartIndex : data.length;

  return mean(data.slice(nextBucketStartIndex, nextNextBucketStartIndex));
};

export const upSample = (values: number[], targetSize: number) => {
  if (!values.length) {
    console.warn('Cannot extend empty array of amplitudes.');
    return values;
  }

  if (values.length > targetSize) {
    console.warn('Requested to extend the waveformData that is longer than the target list size');
    return values;
  }

  if (targetSize === values.length) return values;

  // eslint-disable-next-line prefer-const
  let [bucketSize, remainder] = divMod(targetSize, values.length);
  const result: number[] = [];

  for (let i = 0; i < values.length; i++) {
    const extra = remainder && remainder-- ? 1 : 0;
    result.push(...Array(bucketSize + extra).fill(values[i]));
  }
  return result;
};
