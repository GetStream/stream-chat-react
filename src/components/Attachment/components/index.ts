export * from './DownloadButton';
export * from './FileSizeIndicator';
export * from './ProgressBar';
export * from './PlaybackRateButton';
export * from './PlayButton';
export * from './WaveProgressBar';
