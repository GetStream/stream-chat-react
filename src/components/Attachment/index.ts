export * from './Attachment';
export * from './AttachmentActions';
export * from './AttachmentContainer';
export * from './Audio';
export * from './Card';
export * from './UnsupportedAttachment';
export * from './FileAttachment';
export * from './utils';
