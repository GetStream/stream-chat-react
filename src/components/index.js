export { Avatar } from './Avatar';
export { Message } from './Message';
export { MessageList } from './MessageList';
export { Chat } from './Chat';
export { Channel } from './Channel';
export { ChannelHeader } from './ChannelHeader';
export { MessageInput } from './MessageInput';
export { MessageInputLarge } from './MessageInputLarge';
export { MessageInputFlat } from './MessageInputFlat';
export { MessageInputSmall } from './MessageInputSmall';

export { Attachment } from './Attachment';

export { ChannelList } from './ChannelList';
export { ChannelListMessenger } from './ChannelListMessenger';
export { ChannelListTeam } from './ChannelListTeam';
export { ChannelPreviewCompact } from './ChannelPreviewCompact';
export { ChannelPreviewMessenger } from './ChannelPreviewMessenger';
export { LoadMorePaginator } from './LoadMorePaginator';
export { LoadingChannels } from './LoadingChannels';
export { InfiniteScrollPaginator } from './InfiniteScrollPaginator';
export { LoadingIndicator } from './LoadingIndicator';
export { MessageCommerce } from './MessageCommerce';
export { MessageLivestream } from './MessageLivestream';
export { MessageTeam } from './MessageTeam';
export { MessageSimple } from './MessageSimple';
export { SendButton } from './SendButton';
export { Thread } from './Thread';
export { TypingIndicator } from './TypingIndicator';
export { Window } from './Window';
export { ReactionSelector } from './ReactionSelector';
export { ReactionsList } from './ReactionsList';

export * from '../utils';
export * from '../context';
