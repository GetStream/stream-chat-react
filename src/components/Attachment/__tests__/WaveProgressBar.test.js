import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { downSample, upSample, WaveProgressBar } from '../components';

jest.spyOn(console, 'warn').mockImplementation();
const originalSample = Array.from({ length: 10 }, (_, i) => i);

const PROGRESS_INDICATOR_TEST_ID = 'wave-progress-bar-progress-indicator';
describe('WaveProgressBar', () => {
  describe('component', () => {
    it('is not rendered if waveform data is missing', () => {
      render(<WaveProgressBar seek={jest.fn()} waveformData={[]} />);
      expect(screen.queryByTestId('wave-progress-bar-track')).not.toBeInTheDocument();
    });
    it('is rendered with zero progress by default if waveform data is available', () => {
      const { container } = render(
        <WaveProgressBar amplitudesCount={5} seek={jest.fn()} waveformData={originalSample} />,
      );
      expect(container).toMatchSnapshot();
      expect(screen.queryByTestId(PROGRESS_INDICATOR_TEST_ID)).toBeInTheDocument();
    });
    it('is rendered with highlighted bars with non-zero progress', () => {
      const { container } = render(
        <WaveProgressBar
          amplitudesCount={5}
          progress={20}
          seek={jest.fn()}
          waveformData={originalSample}
        />,
      );
      expect(
        container.querySelectorAll('.str-chat__wave-progress-bar__amplitude-bar--active'),
      ).toHaveLength(1);
      expect(screen.queryByTestId(PROGRESS_INDICATOR_TEST_ID)).toBeInTheDocument();
      expect(screen.queryByTestId(PROGRESS_INDICATOR_TEST_ID)).toHaveStyle('left: 20%');
    });
  });

  describe('amplitude sampling', () => {
    describe('upSample', () => {
      afterEach(jest.restoreAllMocks);
      it('should return original values if target size is smaller than the original sample size', () => {
        expect(upSample(originalSample, 5)).toHaveLength(originalSample.length);
      });

      it('should return original values if the original sample size is empty', () => {
        expect(upSample([], 5)).toHaveLength(0);
      });

      it('should return original values if the original sample size equals the target', () => {
        expect(upSample(originalSample, originalSample.length)).toHaveLength(originalSample.length);
      });

      it('should fill each bucket to reach the target sample size', () => {
        expect(JSON.stringify(upSample(originalSample, 17))).toBe(
          JSON.stringify([0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 8, 9]),
        );
      });
    });

    describe('downSample', () => {
      it('should return original values if target size is greater than the original sample size', () => {
        expect(downSample(originalSample, 20)).toHaveLength(originalSample.length);
      });

      it('should return original values if the original sample size is empty', () => {
        expect(downSample([], 5)).toHaveLength(0);
      });

      it('should return original values if the original sample size equals the target', () => {
        expect(downSample(originalSample, originalSample.length)).toHaveLength(
          originalSample.length,
        );
      });

      it('should return a mean of original values if the target output size is 1', () => {
        expect(JSON.stringify(downSample([10, 2, 6, 10, 3, 4, 8, 0], 1))).toBe(
          JSON.stringify([5.375]),
        );
      });

      it('should fill each bucket to reach the target sample size', () => {
        expect(
          JSON.stringify(
            downSample(
              [10, 2, 6, 10, 3, 4, 8, 1, 10, 0, 6, 10, 3, 4, 8, 1, 2, 6, 10, 3, 8, 10, 0],
              7,
            ),
          ),
        ).toBe(JSON.stringify([10, 2, 10, 0, 8, 10, 0]));
        expect(JSON.stringify(downSample([10, 2, 6, 10, 3, 4, 8, 0], 7))).toBe(
          JSON.stringify([10, 2, 6, 10, 3, 8, 0]),
        );
        expect(JSON.stringify(downSample([10, 2], 2))).toBe(JSON.stringify([10, 2]));
        expect(JSON.stringify(downSample([10, 2, 10], 2))).toBe(JSON.stringify([10, 10]));
      });
    });
  });
});
