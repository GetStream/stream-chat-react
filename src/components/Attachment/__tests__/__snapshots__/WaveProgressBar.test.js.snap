// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaveProgressBar component is rendered with zero progress by default if waveform data is available 1`] = `
<div>
  <div
    class="str-chat__wave-progress-bar__track"
    data-testid="wave-progress-bar-track"
    role="progressbar"
  >
    <div
      class="str-chat__wave-progress-bar__amplitude-bar"
      data-testid="amplitude-bar"
      style="--str-chat__wave-progress-bar__amplitude-bar-height: 0%;"
    />
    <div
      class="str-chat__wave-progress-bar__amplitude-bar"
      data-testid="amplitude-bar"
      style="--str-chat__wave-progress-bar__amplitude-bar-height: 200%;"
    />
    <div
      class="str-chat__wave-progress-bar__amplitude-bar"
      data-testid="amplitude-bar"
      style="--str-chat__wave-progress-bar__amplitude-bar-height: 500%;"
    />
    <div
      class="str-chat__wave-progress-bar__amplitude-bar"
      data-testid="amplitude-bar"
      style="--str-chat__wave-progress-bar__amplitude-bar-height: 700%;"
    />
    <div
      class="str-chat__wave-progress-bar__amplitude-bar"
      data-testid="amplitude-bar"
      style="--str-chat__wave-progress-bar__amplitude-bar-height: 900%;"
    />
    <div
      class="str-chat__wave-progress-bar__progress-indicator"
      data-testid="wave-progress-bar-progress-indicator"
      style="left: 0%;"
    />
  </div>
</div>
`;
